### PR TITLE
[SQL] Add cmdlets for Per DB CMK

### PR DESCRIPTION
**Pull request by viparek**
## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by BethanyZhou** [comment]
Hi @viparek ,

1. please online version for following help docs:
```
 
  Target      : Revalidate-AzSqlDatabaseTransparentDataEncryptionProtector.md
  Description : Online version in the header of the file is incorrect. The corresponding URL should be the following: htt
                ps://learn.microsoft.com/powershell/module/az.sql/revalidate-azsqldatabasetransparentdataencryptionprotec
                tor
  
  Target      : Revalidate-AzSqlInstanceTransparentDataEncryptionProtector.md
  Description : Online version in the header of the file is incorrect. The corresponding URL should be the following: htt
                ps://learn.microsoft.com/powershell/module/az.sql/revalidate-azsqlinstancetransparentdataencryptionprotec
                tor
  
  Target      : Revalidate-AzSqlServerTransparentDataEncryptionProtector.md
  Description : Online version in the header of the file is incorrect. The corresponding URL should be the following: htt
                ps://learn.microsoft.com/powershell/module/az.sql/revalidate-azsqlservertransparentdataencryptionprotecto
                r
  
  Target      : Revert-AzSqlDatabaseTransparentDataEncryptionProtector.md
  Description : Online version in the header of the file is incorrect. The corresponding URL should be the following: htt
                ps://learn.microsoft.com/powershell/module/az.sql/revert-azsqldatabasetransparentdataencryptionprotector

```
2. please re-record all test cases when API version changes.

**Comment by BethanyZhou** [comment]
let me know if the pr is ready to review again by adding label `needs-review`

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceTransparentDataEncryptionProtectorRevalidation", SupportsShouldProcess = true, DefaultParameterSetName = DefaultParameterSet)]
+    [OutputType(typeof(AzureRmSqlManagedInstanceTransparentDataEncryptionProtectorModel))]
+    public class RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector : AzureSqlRmManagedInstanceTransparentDataEncryptionProtectorBase
+    {
+        /// <summary>
+        /// Defines whether it is ok to skip the revalidate Transparent Data Encryption protector request confirmation
+        /// </summary>
+        [Parameter(HelpMessage = "Skip confirmation message for performing the action")]
+        public SwitchParameter Force { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs:20]: -Force here is unnecessary, if you add  SupportsShouldProcess, we will have -Confirm to allow to skip confirmation

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        /// <summary>
+        /// Gets or sets whether or not to run this cmdlet in the background as a job
+        /// </summary>
+        [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+        public SwitchParameter AsJob { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs:26]: Could you confirm your operation whether is a LRO? only LRO needs AsJob.

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    /// Defines the Revalidate-AzSqlServerTransparentDataEncryptionProtector cmdlet
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerTransparentDataEncryptionProtectorRevalidation", SupportsShouldProcess = true), OutputType(typeof(AzureSqlServerTransparentDataEncryptionProtectorModel))]
+    public class RevalidateAzureSqlServerTransparentDataEncryptionProtector : AzureSqlServerTransparentDataEncryptionProtectorCmdletBase
+    {
+        /// <summary>
+        /// Defines whether it is ok to skip the revalidate Transparent Data Encryption protector request confirmation
+        /// </summary>
+        [Parameter(HelpMessage = "Skip confirmation message for performing the action")]
+        public SwitchParameter Force { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs:32]: please check -Force and AsJob

File: src/Sql/Sql/Database Backup/Model/AzureSqlDeletedDatabaseBackupModel.cs [hunk 4]
```diff
@@ -80,5 +81,10 @@ public class AzureSqlDeletedDatabaseBackupModel
         /// Gets or sets the resource ID
         /// </summary>
         public string ResourceId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the list of database AKV keys.
+        /// </summary>
+        public List<string> Keys { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database Backup/Model/AzureSqlDeletedDatabaseBackupModel.cs:88]: It's recommend using string array as the type of Keys

File: src/Sql/Sql/Database/Cmdlet/NewAzureSqlDatabase.cs [hunk 5]
```diff
@@ -248,6 +250,41 @@ public class NewAzureSqlDatabase : AzureSqlDatabaseCmdletBase<AzureSqlDatabaseCr
...
+        [Parameter(Mandatory = false,
+            HelpMessage = "The encryption protector key for SQL Database.")]
+        public string EncryptionProtector { get; set; }
+
+        /// <summary>
+        /// List of user assigned managed identities
+        /// </summary>
+        [Parameter(Mandatory = false,
+            HelpMessage = "The list of user assigned identity for the SQL Database.")]
+        public List<string> UserAssignedIdentityId { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/NewAzureSqlDatabase.cs:272]: Please use string array to type of UserAssignedIdentityId, see https://github.com/Azure/azure-powershell/blob/main/documentation/development-docs/design-guidelines/parameter-best-practices.md#array-vs-enumerable-types

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

**Review by strehan1993 — DISMISSED** [review]

File: src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            HelpMessage = "The name of the Azure SQL Database to revalidate.")]
+        [Alias("Name")]
+        [ValidateNotNullOrEmpty]
+        public string DatabaseName { get; set; }
+
+        /// <summary>
+        /// Defines whether it is ok to skip the revalidate Transparent Data Encryption protector request confirmation
+        /// </summary>
+        [Parameter(HelpMessage = "Skip confirmation message for performing the action")]
+        public SwitchParameter Force { get; set; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs:25]: Per my knowledge, validating doesn't change data/resource am I correct? Force is needed if we wan to skip confirmation to change resource.  Please check all cmdlets.

File: src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
+﻿using Microsoft.Azure.Commands.Sql.Database.Model;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Sql.Database.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet for Revalidate Azure Sql Database encryption protector
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlDatabaseTransparentDataEncryptionProtectorRevalidation", SupportsShouldProcess = true), OutputType(typeof(AzureSqlDatabaseModel))]
+    public class RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector : AzureSqlDatabaseCmdletBase<AzureSqlDatabaseModel>
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs:10]: Change the class name and file name to match invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation

File: src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        ///  Defines whether the Cmdlets will output the model object at the end of its execution
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        public SwitchParameter PassThru { get; set; }
+
+        /// <summary>
+        /// Returns true if the model object that was constructed by this cmdlet should be written out.
+        /// </summary>
+        /// <returns>True if the model object should be written out; false otherwise.</returns>
+        protected override bool WriteResult() { return PassThru; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/RevalidateAzureSqlDatabaseTransparentDataEncryptionProtector.cs:43]:
```suggestion
        protected override bool WriteResult() { return PassThru.IsPresent; }
```

File: src/Sql/Sql/Database/Cmdlet/RevertAzureSqlDatabaseTransparentDataEncryptionProtector.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.Commands.Sql.Database.Cmdlet
+{
+    /// <summary>
+    /// Cmdlet for Revert Azure Sql Database encryption protector
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlDatabaseTransparentDataEncryptionProtectorRevert", SupportsShouldProcess = true), OutputType(typeof(AzureSqlDatabaseModel))]
+    public class RevertAzureSqlDatabaseTransparentDataEncryptionProtector : AzureSqlDatabaseCmdletBase<AzureSqlDatabaseModel>
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/RevertAzureSqlDatabaseTransparentDataEncryptionProtector.cs:13]: Change the class name and file name to match invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevert

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Sql.TransparentDataEncryption.Cmdlet
+{
+    /// <summary>
+    /// Defines the Revalidate-AzSqlInstanceTransparentDataEncryptionProtector cmdlet
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlInstanceTransparentDataEncryptionProtectorRevalidation", SupportsShouldProcess = true, DefaultParameterSetName = DefaultParameterSet)]
+    [OutputType(typeof(AzureRmSqlManagedInstanceTransparentDataEncryptionProtectorModel))]
+    public class RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector : AzureSqlRmManagedInstanceTransparentDataEncryptionProtectorBase
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs:14]: Change the class name and file name to match invoke-AzSqlInstanceTransparentDataEncryptionProtectorRevalidation

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Sql.TransparentDataEncryption.Cmdlet
+{
+    /// <summary>
+    /// Defines the Revalidate-AzSqlServerTransparentDataEncryptionProtector cmdlet
+    /// </summary>
+    [Cmdlet("Invoke", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "SqlServerTransparentDataEncryptionProtectorRevalidation", SupportsShouldProcess = true), OutputType(typeof(AzureSqlServerTransparentDataEncryptionProtectorModel))]
+    public class RevalidateAzureSqlServerTransparentDataEncryptionProtector : AzureSqlServerTransparentDataEncryptionProtectorCmdletBase
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs:26]: Change class name and file name to match invoke-azSqlServerTransparentDataEncryptionProtectorRevalidation

File: src/Sql/Sql/Database/Cmdlet/RevertAzureSqlDatabaseTransparentDataEncryptionProtector.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        ///  Defines whether the Cmdlets will output the model object at the end of its execution
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        public SwitchParameter PassThru { get; set; }
+
+        /// <summary>
+        /// Returns true if the model object that was constructed by this cmdlet should be written out.
+        /// </summary>
+        /// <returns>True if the model object should be written out; false otherwise.</returns>
+        protected override bool WriteResult() { return PassThru; }
```
Comment (BethanyZhou) [inline src/Sql/Sql/Database/Cmdlet/RevertAzureSqlDatabaseTransparentDataEncryptionProtector.cs:46]:
```suggestion
        protected override bool WriteResult() { return PassThru.IsPresent; }
```

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        /// <summary>
+        /// Gets or sets whether or not to run this cmdlet in the background as a job
+        /// </summary>
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureRmSqlManagedInstanceTransparentDataEncryptionProtector.cs:18]: Add PassThru to keep consistent with invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation

File: src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        /// <summary>
```
Comment (BethanyZhou) [inline src/Sql/Sql/TransparentDataEncryption/Cmdlet/RevalidateAzureSqlServerTransparentDataEncryptionProtector.cs:28]: Add PassThru to keep consistent with invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

**Comment by BethanyZhou** [comment]
Please resolve conflicts first and fix a test case named TestDatabaseUpdateWithPerDBCMK, which is added by you. @viparek

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Sql/Sql.Sdk/Microsoft.Azure.Management.Sql.csproj [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (BethanyZhou) [inline src/Sql/Sql.Sdk/Microsoft.Azure.Management.Sql.csproj:4]: This file should not be checked in.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Sql/Sql/Az.Sql.psd1 [hunk 16]
```diff
@@ -293,7 +293,11 @@ CmdletsToExport = 'Get-AzSqlDatabaseTransparentDataEncryption',
                'Update-AzSqlServerAdvancedThreatProtectionSetting', 
                'Update-AzSqlInstanceDatabaseAdvancedThreatProtectionSetting', 
                'Update-AzSqlInstanceAdvancedThreatProtectionSetting', 
-               'Get-AzSqlInstanceDtc', 'Set-AzSqlInstanceDtc'
+               'Get-AzSqlInstanceDtc', 'Set-AzSqlInstanceDtc',
+			   'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation',
+			   'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevert',
+			   'Invoke-AzSqlServerTransparentDataEncryptionProtectorRevalidation',
+			   'Invoke-AzSqlInstanceTransparentDataEncryptionProtectorRevalidation'
```
Comment (BethanyZhou) [inline src/Sql/Sql/Az.Sql.psd1:300]:
```suggestion
			   'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation',
			   'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevert',
			   'Invoke-AzSqlServerTransparentDataEncryptionProtectorRevalidation',
			   'Invoke-AzSqlInstanceTransparentDataEncryptionProtectorRevalidation'
```
```suggestion
               'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevalidation',
               'Invoke-AzSqlDatabaseTransparentDataEncryptionProtectorRevert',
               'Invoke-AzSqlServerTransparentDataEncryptionProtectorRevalidation',
               'Invoke-AzSqlInstanceTransparentDataEncryptionProtectorRevalidation'
```

**Review by BethanyZhou — APPROVED** [review]
Looks good to me

**Pull request merged**